1.0-dev: fully disallow conversion between non `Interval` types

OlivierHnt opened the pull request:
## Context

In the recent PR #555 the conversion/promotion behaviour between non `Interval` types was discussed.
The agreed plan for now is to **not allow generic code to conflate** `Number` and `Interval`. Instead, this should be done in a case-by-case fashion.

However, in the current state of the 1.0-dev branch, `Interval{T}` is defined as a constructor for a wide range of inputs, in particular for `Interval(::Real)`. While convenient, there is a conflict here because in Julia `(::Type{T})(a) where {T<:Number}` is the default behaviour for conversion.
So, defining a method `Interval{T}(a::Real) where {T} = Interval{T}(a, a)` implies that `convert(Interval{Float64}, 1)` or `convert(Interval{Float64}, pi)` works (meanwhile `promote_type(Interval{Float64}, Int)` and `promote_type(Interval{Float64}, Irrational{:pi})` error on purpose). In my eyes this feels like a curious inconsistency.

N.B.: the aforementioned conversion process always return a **correct** interval. Thus, this discussion is not about the validity of the conversion process, but rather that conversion in the wild may have undesirable side-effects with generic codes.

## Content of the PR

This PR fully (I think) prevent unwanted conversion by reserving the constructor `Interval{T}` to a very limited set of signatures. After this PR, the **only allowed conversion** is `Interval{T}(::Interval)` which coincides with the **only allowed promotion** `promote_type(::Interval, ::Interval)`.

All the remaining constructors `Interval{T}(...)` have been modified into methods `interval(::Type{T}, ...)`. I believe that this is a sensible choice since constructing an interval should always abide by the IEEE standard (except perhaps internally for performance reasons). Also, this unifies better the constructors by reinforcing the usage of the function `interval`.

Pros:
- enforce the splitting between `Interval` and `Number` by preventing all conversion except between `Interval`s.
- still provide a way to reintroduce conversion later on if we change our mind.
- I decided to only allow the unsafe constructor `Interval{T}(a, b)` since I like that it is a little cumbersome to use. In this PR, `interval` becomes the go-to constructor.

Cons:
- preventing conversion makes using `Interval` less flexible and hinders writing generic code, see https://github.com/JuliaIntervals/IntervalArithmetic.jl/pull/271#issuecomment-1021384615

## Other comments

1. Many files needed to be slightly changed. The general gist of the modifications are as follows. Calls like `Interval{T}(a)` have been generally changed to `Interval{T}(a, a)` so keeping the exact behaviour as before. Except for `+`, `-` and `*` (cf src/intervals/arithmetic/basic.jl) where I use the safe constructor `interval(T, a)`. As a consequence `( 1..2 ) + Inf` now returns an empty interval (before it would return `Interval(Inf, Inf)`). This is a take on fixing the issue #535.

2. I needed to change the syntax of the display for `:full` because it must be parseable Julia code. So instead of displaying `Interval(..., ...)` we now have `Interval{T}(..., ...)`.

3. There were inconsistencies with the behaviour of `convert(::Type{Interval{T}}, ::Complex)` and `Interval{T}(::Complex)` which would return a `Complex{Interval{T}}`. In Base (cf. base/complex.jl) the convention is that `(::Type{T})(::Complex) where {T<:Real}` either errors or returns an instance of `T`. This PR fixes this issue.

Example:

```
julia> convert(Interval{Float64}, 1+0im)  # Consistent with the "no wild conversion policy" between `Number` and `Interval`
ERROR: MethodError: no method matching Interval{Float64}(::Int64)
...

julia> x = (1..1) + (0..0)*im
[1, 1] + [0, 0]im

julia> convert(Interval{Float64}, x)  # `imag(x) == 0`. Conversion to a `Real` as in Base, see e.g. convert(Float64, 1.0+0.0im)
[1.0, 1.0]

julia> y = (1..1) + (1..1)*im
[1, 1] + [1, 1]im

julia> convert(Interval{Float64}, y)  # `imag(y) != 0`. Cannot convert to a `Real` as in Base, see e.g. convert(Float64, 1.0+1.0im)
ERROR: InexactError: Interval(Interval{Float64}(1.0, 1.0) + Interval{Float64}(1.0, 1.0)im)
...

julia> convert(Interval{Float64}, interval(Float32, π)+(0..0)*im)  # Conversion between different `Interval` types is allowed
[3.14159, 3.1416]
```

4. Note that now expressions such as `1..1 + 0im` fails. It is unfortunate but it can easily be fixed in a different PR focusing on improving support for complex numbers.

lbenet commented on test/interval_tests/numeric.jl:
```diff
@@ -9,138 +9,138 @@ using Test
-    @test b/a ≛ Interval(/(b.lo, a.hi, RoundDown), /(b.hi, a.lo, RoundUp))
-    @test a/c ≛ Interval(0.025, 4.4)
-    @test c/4.0 ≛ Interval(6.25e-02, 1e+00)
+    @test interval(1//4,1//2) + interval(2//3) ≛ interval(11//12, 7//6)
+    @test_broken interval(1//4,1//2) - interval(2//3) ≛ interval(-5//12, -1//6)
```
Replace by:
```julia
@test interval(Rational{Int}, 1//4,1//2) + interval(Rational{Int}, 2//3) ≛ interval(Rational{Int}, 11//12, 7//6)
@test interval(1//4,1//2) - interval(2//3) ≛ interval(-5//12, -1//6)
```

lbenet reviewed (commented):
Thanks a lot for this PR, and sorry for the long to review it. Thank you very much also for describing the context and the essential contents; that truly helped understanding it! I agree that limiting the use of `Interval{T}` is useful and may prevent errors/problems. So, in general, it LGTM. 

I have left a bunch of comments/suggestions that are hopefully useful. In my opinion we should avoid the possibility of creating `Interval{Int}` (as it is now in 1.0-dev, I think), but allowing `Interval{Rational{Int}}`. Also, there are some inconsistencies with `Interval{BigFloat}` and/or using `Irrational`s, which may already be present in 1.0-dev.

OlivierHnt commented on test/interval_tests/construction.jl:
```diff
@@ -5,47 +5,46 @@ using Test
 
-    @test_broken ℯ..big(4) ≛ hull(Interval{BigFloat}(π), interval(4))
-    @test π..big(4) ≛ hull(Interval{BigFloat}(π), interval(4))
+    @test_broken ℯ..big(4) ≛ hull(interval(BigFloat, π), interval(4))
+    @test π..big(4) ≛ hull(interval(BigFloat, π), interval(4))
```
Cf. https://github.com/JuliaIntervals/IntervalArithmetic.jl/pull/558#discussion_r1143266710

OlivierHnt commented on src/intervals/construction.jl:
```diff
@@ -64,71 +64,68 @@ struct Interval{T} <: Real
+    interval(promote_type(T, S), a, b)
+interval(a::T, b::S) where {T<:AbstractFloat, S} = interval(promote_type(T, S), a, b)
+interval(a::T, b::S) where {T, S<:AbstractFloat} = interval(promote_type(T, S), a, b)
+interval(a::T, b::S) where {T, S} = interval(promote_type(default_bound(), T, S), a, b)
```
I kept this because it was the behaviour before the PR.

When we restrict the parametric type to `Union{Rational,AbstractFloat}`, I think we should actually prevent this from happening; what do you think?

OlivierHnt commented on src/intervals/construction.jl:
```diff
@@ -64,71 +64,68 @@ struct Interval{T} <: Real
+interval(a::T, b::S) where {T, S} = interval(promote_type(default_bound(), T, S), a, b)
+
+# Real: `is_valid_interval(a, a) != true`
+interval(::Type{T}, a::Real) where {T} = interval(T, a, a)
```
Good catch 😅. I could special case for `T<:Int` and `T<:Irrational`, but this won't really be a long term fix because the root of the problem is that the parametric type of `Interval` can be `Any`.
So either we wait for the PR when we restrict to `Union{Rational, AbstractFloat}`; unless it seems that this should be addressed in this PR.

At least the problematic behaviour `Interval{Irrational{:π}}(π, π)` looks harmless since any operations with it appears to throw an error.

Kolaru replied:
I think this is fine for this PR, since it only matters for explicit input from user.

OlivierHnt commented on src/decorations/intervals.jl:
```diff
@@ -33,9 +33,25 @@ end
-    is_valid_interval(a, b) || return DecoratedInterval(Interval(a,b), ill)
-    return DecoratedInterval(Interval(a,b), d)
+function DecoratedInterval(a::T, b::S, d::DECORATION) where {T<:Real, S<:Real}
+    BoundsType = promote_type(default_bound(), T, S)
```
I think this was also the behaviour before this PR.
Perhaps we want to tackle once we restrict the bounds type of `Interval` to `Union{Rational, AbstractFloat}`, see also https://github.com/JuliaIntervals/IntervalArithmetic.jl/pull/558#discussion_r1143278922

OlivierHnt commented:
@lbenet Thank you for the thorough feedback! I fixed the minor things and left some other comments.
What seems to be conflicting are some odd behaviours relating to the parametric type restriction of `Interval`.

I left most behaviour (even the odd ones) untouched in this PR because I initially wanted to make an other PR where I restrict the bounds type of an interval to `Union{Rational, AbstractFloat}`. I thought this would make changes more traceable.
That being said, I am also willing to implement this type restriction in this PR if you think that it will be fine.

Kolaru commented on src/intervals/construction.jl:
```diff
@@ -64,71 +64,68 @@ struct Interval{T} <: Real
-
-#= Complex =#
-Interval(x::Complex) = Interval(real(x)) + im*Interval(imag(x))
+Interval{T}(a, b) where {T} = Interval{T}(T(a, RoundDown), T(b, RoundUp))
```
We can define `BigFloat(::Irrational{:ℯ}, ::RoundingMode)` ourselves as a hotfix, the rest seems to be a Base problem indeed. To do it properly (i.e. tightlty), the macro defining those kind of constant probably needs to be adapted.

Kolaru reviewed (changes requested):
Thanks very much this is very nice work !

Two more general comments/questions :
1. For the ITF tests did you change them manually or did you modify the upstream package that generate them ? (the latter would be better)
2. Do you think it is possible to come up with some test that show that the disallowed conversions are disallowed indeed ? I didn't see any new tests, so I'm wondering.

OlivierHnt commented:
@Kolaru

1. I modified it manually; I did not know about [ITF1788.jl](https://github.com/JuliaIntervals/ITF1788.jl) until I read your message. Mhmm so how should we go about this? Create a "tests-1.0-dev" branch on that repo?

2. I already modified the tests to show the absence of conversion, see https://github.com/OlivierHnt/IntervalArithmetic.jl/blob/09ac4da05bb4837a7306710251149017dfa50f75/test/interval_tests/construction.jl#L58-L64. Do you have other tests in mind?

OlivierHnt commented on src/decorations/intervals.jl:
```diff
@@ -33,9 +33,25 @@ end
-function DecoratedInterval(a::Real, b::Real, d::DECORATION)
-    is_valid_interval(a, b) || return DecoratedInterval(Interval(a,b), ill)
-    return DecoratedInterval(Interval(a,b), d)
+function DecoratedInterval(a::T, b::S, d::DECORATION) where {T<:Real, S<:Real}
```
Good catch. I went for the name `getnumtype` (not exported) for consistency with `numtype`

OlivierHnt commented:
Friendly bump 🙂

lbenet reviewed (approved):
I'm in favor of merging this PR as it is. Once said that, it would be useful to have an issue where the items not covered in this PR, but that arose in the review, could be listed.
cc @Kolaru

Kolaru commented:
Yes sorry, I forgot to come back to it.

> 1. I modified it manually; I did not know about [ITF1788.jl](https://github.com/JuliaIntervals/ITF1788.jl) until I read your message. Mhmm so how should we go about this? Create a "tests-1.0-dev" branch on that repo?

That seems reasonnable. I don't know how @lucaferranti usually did deal with changing tests.

> 2. I already modified the tests to show the absence of conversion, see https://github.com/OlivierHnt/IntervalArithmetic.jl/blob/09ac4da05bb4837a7306710251149017dfa50f75/test/interval_tests/construction.jl#L58-L64. Do you have other tests in mind?

Thanks for the pointer, I missed that during my review. It seems to do be sufficient.

So it looks good to me, thanks a lot for your work @OlivierHnt !
